src/lib/sw.js

```diff
@@ -1,31 +1,86 @@
+    // foo/index.html will not load foo/).
+    if (request.url.endsWith("/")) {
+      return request.url + "index.html";
+    }
+    return request;
+  },
+};
+
 import manifest from "cache-manifest";
 
 importScripts(
   "https://storage.googleapis.com/workbox-cdn/releases/4.3.1/workbox-sw.js",
 );
 
-console.log("Got cache manifest", manifest);
+workbox.precaching.precacheAndRoute(manifest);
 
-workbox.precaching.precache(["/offline/"]);
+/**
+ * Match "/foo-bar/ "and "/foo-bar/as/many/of-these-as-you-like/" (with optional trailing
+ * "index.html"), normal page nodes for web.dev.
+ */
+const contentPageRe = new RegExp("/([\\w-]+/)*(|index.html)$");
+
+/**
+ * Match "/foo-bar" and "/foo-bar/as-many/but/no/trailing/slash" (but not "/foo/bar/index.html").
+ */
+const untrailedContentPageRe = new RegExp("(/[\\w-]+)+$");
+
+/**
+ * Send normal nodes to cache first.
+ */
+workbox.routing.registerRoute(
+  contentPageRe,
+  new workbox.strategies.StaleWhileRevalidate({
+    plugins: [normalizeIndexCacheKeyPlugin],
+  }),
+);
 
 /**
- * Match /foo-bar/ and "/foo-bar/as/many/of-these-as-you-like/".
- *
- * TODO(samthor): Handle /index.html suffix via cacheKey and the Express-generated 302 from "/foo"
- * to "/foo/".
- *
- * TODO(samthor): Workbox is agressive and will also match "other-domain.com/foo/", include domain
- * restriction.
+ * Cache images that aren't included in the original manifest, such as author profiles.
  */
 workbox.routing.registerRoute(
-  new RegExp("/([\\w-]+/)*$"),
+  new RegExp("/images/.*"),
   new workbox.strategies.StaleWhileRevalidate(),
 );
 
-workbox.routing.setCatchHandler(({event}) => {
+/**
+ * Untrailed requests are network-only, but with a fallback to redirecting to the same page with a
+ * trailing slash.
+ */
+self.addEventListener("fetch", (event) => {
+  const u = new URL(event.request.url);
+  if (
+    !untrailedContentPageRe.exec(event.request.url) ||
+    self.location.host !== u.host
+  ) {
+    return;
+  }
+
+  // Go to the network first, looking for interesting responses from the server (e.g., "/subscribe"
+  // redirects to a form).
+  const p = fetch(event.request).catch(() => {
+    // Otherwise, just assume that the content is actually at the trailed location.
+    const headers = new Headers();
+    headers.append("Location", event.request.url + "/");
+    const redirectResponse = new Response("", {
+      status: 301,
+      headers,
+    });
+    return redirectResponse;
+  });
+
+  event.respondWith(p);
+});
+
+workbox.routing.setCatchHandler(async ({event}) => {
   if (event.request.destination === "document") {
     // TODO(samthor): Annotate this page so the client knows it's being displayed because the user
     // is offline, and force it to rerequest when navigator.onLine is true.
-    return caches.match("/offline/");
+    // We have to ignoreSearch because Workbox adds __WB_REVISION__ to all its cached contents.
+    return caches.match("/offline/index.html", {ignoreSearch: true});
```

Comment:
You could use `workbox.precaching.getCacheKeyForURL("/offline/index.html")` as an alternative to setting `ignoreSearch`. (That has the advantage of always returning the version of `index.html` corresponding to the currently active service worker, which admittedly isn't likely to be a big deal for something like an offline page.)

Reply:
I was going to tell you to write this in the docs then I realised that it's literally right there. 🤦‍♂ Thanks.